Add OutputType to `Get-Error` cmdlet and preserve original typenames

SteveL-MSFT opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Cmdlet was missing declaring OutputType which is `ErrorRecord#PSExtendedError` and `Exception#PSExtendedError`.  Added OutputType attribute.  Cmdlet adds the `PSExtendedError` typename and removes `Exception` and `ErrorRecord` typenames so that the formatting is used.  The formatter then removes `PSExtendedError` and puts back the original typename so that `$Error` should be the same before calling `Get-Error`.  While testing, had to make some changes to how InvocationInfo is retrieved so that ParseException which contains a nested ErrorRecord which as InvocationInfo is handled correctly.  Combined Exception and ErrorRecord formatter into one.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

SteveL-MSFT committed:
add OutputType to cmdlet

felixfbecker commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -15,6 +15,7 @@ namespace Microsoft.PowerShell.Commands
     [Cmdlet(VerbsCommon.Get, "Error",
         HelpUri = "https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/get-error?view=powershell-7&WT.mc_id=ps-gethelp",
         DefaultParameterSetName = NewestParameterSetName)]
+    [OutputType(typeof(ErrorRecord), typeof(Exception))]
```
Not sure if this is valid
```suggestion
    [OutputType('PSExtendedError', typeof(ErrorRecord), typeof(Exception))]
```

vexx32 replied:
This is C# so you need double quotes, but otherwise I believe that's perfectly valid :)

SteveL-MSFT replied:
`OutputType()` takes types, not strings.  `PSExtendedError` is a fake type applied to ErrorRecords and Exceptions.

vexx32 replied:
![image](https://user-images.githubusercontent.com/32407840/67248406-61ed6200-f432-11e9-8266-4a769238ffeb.png)

It'll do both, though I'm not sure if PSTypeNames or fake types like that "work". 🙂

SteveL-MSFT replied:
Hmm, I guess I can just say it returns PSExtendedError since I explicitly remove the ErrorRecord and Exception typenames so they don't get priority in rendering.  Will update.

SteveL-MSFT committed:
Change OutputType to just PSExtendedError

PoshChan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -15,6 +15,7 @@ namespace Microsoft.PowerShell.Commands
     [Cmdlet(VerbsCommon.Get, "Error",
```
@SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-windows`
Invoke-WebRequest tests.Denial of service.Image Parsing
```powershell
Expected the actual value to be greater than 10, but got 6.63326583260782.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\WebCmdlets.Tests.ps1: line 1897
1897:             $pathologicalRatio | Should -BeGreaterThan 10
```

PoshChan replied:
@SteveL-MSFT, your last commit had 22 failures in `PowerShell-CI-static-analysis`
(These are 5 of the failures)

Verify Markdown Links.Verify links in /home/vsts/work/1/s/test/tools/CodeCoverageAutomation/README.md.https://github.com/csMACnz/coveralls.net/releases/download/0.7.0/coveralls.net.0.7.0.nupkg should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v6.2.3/PowerShell-6.2.3-win-x64.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0-preview.4/PowerShell-7.0.0-preview.4-win-x64.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v6.2.3/PowerShell-6.2.3-win-x86.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0-preview.4/PowerShell-7.0.0-preview.4-win-x86.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

PoshChan replied:
@SteveL-MSFT, your last commit had 31 failures in `PowerShell-CI-static-analysis`
(These are 5 of the failures)

Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v6.2.3/PowerShell-6.2.3-win-x64.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0-preview.4/PowerShell-7.0.0-preview.4-win-x64.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v6.2.3/PowerShell-6.2.3-win-x86.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0-preview.4/PowerShell-7.0.0-preview.4-win-x86.msi should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v6.2.3/powershell_6.2.3-1.ubuntu.18.04_amd64.deb should work
```powershell
retry of URL failed with error: Resource temporarily unavailable
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

SteveL-MSFT commented:
@PoshChan please retry windows

PoshChan commented:
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Windows`

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -15,6 +15,7 @@ namespace Microsoft.PowerShell.Commands
     [Cmdlet(VerbsCommon.Get, "Error",
         HelpUri = "https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/get-error?view=powershell-7&WT.mc_id=ps-gethelp",
         DefaultParameterSetName = NewestParameterSetName)]
+    [OutputType("PSExtendedError")]
```
```powershell
 (Get-Error).Gettype()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     ErrorRecord                              System.Object
```

```powershell
 $a.psobject.TypeNames
PSExtendedError
PSExtendedError
PSExtendedError
System.Object
```
I wonder to see the difference.
Also for second output I'd expect to see `ErrorRecord` too, not only 3 (!) fake types.

SteveL-MSFT replied:
The multiple is due to everytime you `Get-Error()`, it inserts it.  I'll fix that in this PR.

felixfbecker replied:
Interesting, I hadn't thought about that this would mutate the object. So then, after you called `Get-Error` once, `$error` would suddenly also show the same formatting, right? With no way to get the original formatting again other than manually mutating again? That's a little surprising.

Is there no way the format could just be defined for the `ErrorRecord` type, but as a `CustomControl` that is not used by default but only in certain contexts?

vexx32 replied:
It would probably be prudent to simply `Copy()` the PSObject wrapper and maintain the internals, so you only add the PSTypeName to a fresh instance.

SteveL-MSFT replied:
`Copy()` copies the refs so it doesn't help here as I'm still manipulating the original object.

SteveL-MSFT committed:
have formatter preserve the original types
get-error outputtype is errorrecord and exception
fix formatting that wasn't handling parse exceptions correctly
update tests to check typenames

SteveL-MSFT committed:
change OutputType back to PSExtendedError, add test for it

SteveL-MSFT committed:
address CodeFactor issues

SteveL-MSFT commented:
@PoshChan please retry static

PoshChan commented:
@SteveL-MSFT, successfully started retry of `PowerShell-CI-static-analysis`

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -75,11 +76,27 @@ protected override void ProcessRecord()
-                obj.TypeNames.Remove("System.Exception");
+                if (obj.TypeNames.Contains("System.Management.Automation.ErrorRecord"))
+                {
+                    if (!obj.TypeNames.Contains("System.Management.Automation.ErrorRecord#PSExtendedError"))
```
it's probably ok, but since you use this string a bit, you might turn it into a const

SteveL-MSFT replied:
Yeah, I'll turn into const.

JamesWTruher commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -913,6 +913,16 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
+                                $_.PSObject.TypeNames.Add('System.Management.Automation.ErrorRecord')
+                                $null = $_.PSObject.TypeNames.Remove('System.Management.Automation.ErrorRecord#PSExtendedError')
+                            }
+                            elseif ($_.PSObject.TypeNames.Contains('System.Exception#PSExtendedError')) {
```
elseif? Is it possible for typenames to contain _both_ Exception#PSExtendedError _and_ ErrorRecord#PSExtendedError

SteveL-MSFT replied:
Based on current code, it is not possible as Get-Error only adds one or (else) the other.

JamesWTruher commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -1050,7 +1060,7 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
                                         $newline = [Environment]::Newline
 
-                                        if ($myinv -and $myinv.ScriptName -or $myinv.ScriptLineNumber -gt 1 -or $_.CategoryInfo.Category -eq 'ParserError') {
+                                        if ($myinv -and $myinv.ScriptName -or $myinv.ScriptLineNumber -gt 1 -or $err.CategoryInfo.Category -eq 'ParserError') {
```
just ParserError? not `@('ParserError') -contains $err.CategoryInfo.Category`?

SteveL-MSFT replied:
From my experimentation, only ParserError needs to be covered here as it provides additional context on the location within the line of the issue.

JamesWTruher commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -1180,23 +1197,23 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
                                         if ($null -ne $errorCategoryMsg)
                                         {
-                                            $indentString = '+ CategoryInfo          : ' + $_.ErrorCategory_Message
+                                            $indentString = '+ CategoryInfo          : ' + $err.ErrorCategory_Message
```
hardcoded tab?

SteveL-MSFT replied:
This is the existing NormalView code for the spaces.  Yeah, it's hardcoded to match FullyQualifiedErrorId spacing.

JamesWTruher commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -1180,23 +1197,23 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
                                         $posmsg += ""`n"" + $indentString
 
-                                        $indentString = ""+ FullyQualifiedErrorId : "" + $_.FullyQualifiedErrorId
+                                        $indentString = ""+ FullyQualifiedErrorId : "" + $err.FullyQualifiedErrorId
```
not sure about this one, it looks like it should be:
```suggestion
                                        $indentString += "" FullyQualifiedErrorId : "" + $err.FullyQualifiedErrorId
```

SteveL-MSFT replied:
This is the existing code that literally adds a `+` to the left side of the output.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Error.Tests.ps1:
```diff
@@ -26,6 +26,17 @@ Describe 'Get-Error tests' -Tag CI {
         $out | Should -BeLikeExactly '*InnerException*'
+
+        $err = Get-Error
+        $err.GetType().Name | Should -BeExactly 'ErrorRecord'
```
it would be good to have the same tests for an exception, if possible

SteveL-MSFT replied:
Can't find a way to have a pure Exception in the test script.  If I type `1/d` interactively, I get a ParseException, but if I use `Invoke-Expression` in the Pester test, that cmdlet wraps it in an ErrorRecord.

iSazonov replied:
Maybe `Should -Throw -PassThru` helps?

SteveL-MSFT replied:
That doesn't help as it's already wrapped in an ErrorRecord by `Invoke-Expression`.  I tried using `pwsh -c`, but that ends up wrapping it as an ErrorRecord as well.  Only if I do `1/d` interactively in the console do I get a ParseException that isn't wrapped.

JamesWTruher reviewed (changes requested):
I think you have a typo in there

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -75,11 +76,27 @@ protected override void ProcessRecord()
+                        obj.TypeNames.Insert(0, "System.Management.Automation.ErrorRecord#PSExtendedError");
+
+                        // Need to remove so this rendering doesn't take precedence
+                        obj.TypeNames.Remove("System.Management.Automation.ErrorRecord");
```
Why isn't this a problem for Get-Process where we add "fake" types too?

vexx32 replied:
Don't think this is an issue, but potentially it might be because the error types tend to take precedence in the formatters explicitly (same reason why you need to `$error | Format-List -Force` and without `-Force` it simply gives the standard error view)

SteveL-MSFT replied:
Correct.  There is a whole concept of "Out of Band" formatting which takes precedence.  This is why when you do a recursive directory and you encounter an error, that error is in red and not part of the table output.  However, the side effect of this is these typenames need to be removed otherwise they will be formatted before my formatter gets a chance.

SteveL-MSFT commented:
@JamesWTruher can you point out where exactly is the typo you are referring to?

SteveL-MSFT commented:
@JamesWTruher can you point out where exactly is the typo you are referring to?

SteveL-MSFT committed:
added Exception test and address Jim's feedback

iSazonov commented:
@SteveL-MSFT Do you want to get approve from anybody?

SteveL-MSFT commented:
@JamesWTruher can you review?

iSazonov commented:
Up.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Error.Tests.ps1:
```diff
@@ -26,6 +26,17 @@ Describe 'Get-Error tests' -Tag CI {
+        $err.PSObject.TypeNames | Should -Not -Contain 'System.Management.Automation.ErrorRecord'
+        $err.PSObject.TypeNames | Should -Contain 'System.Management.Automation.ErrorRecord#PSExtendedError'
+
+        # need to exercise the formatter
```
you might elaborate on this comment. Casual inspection looks like you're just testing the same thing twice.

SteveL-MSFT replied:
Will expand comment

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Error.Tests.ps1:
```diff
@@ -80,14 +91,21 @@ Describe 'Get-Error tests' -Tag CI {
+        $err.PSObject.TypeNames | Should -Not -Contain 'System.Exception'
+        $err.PSObject.TypeNames | Should -Contain 'System.Exception#PSExtendedError'
+
+        # need to exercise the formatter
```
and here

JamesWTruher reviewed (commented):
just a couple questions

SteveL-MSFT committed:
Address Jim's feedback

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-Error.cs:
```diff
@@ -75,11 +78,27 @@ protected override void ProcessRecord()
-                // Remove some types so they don't get rendered by those formats
-                obj.TypeNames.Remove("System.Management.Automation.ErrorRecord");
-                obj.TypeNames.Remove("System.Exception");
+                if (obj.TypeNames.Contains("System.Management.Automation.ErrorRecord"))
```
Will remove `else` to handle this case

SteveL-MSFT commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -1050,7 +1060,7 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
                                         $newline = [Environment]::Newline
 
-                                        if ($myinv -and $myinv.ScriptName -or $myinv.ScriptLineNumber -gt 1 -or $_.CategoryInfo.Category -eq 'ParserError') {
+                                        if ($myinv -and $myinv.ScriptName -or $myinv.ScriptLineNumber -gt 1 -or $err.CategoryInfo.Category -eq 'ParserError') {
```
$err is in script scope so not passed into this internal function but is defined before this function is called
